File: leetcode/450.delete-node-in-bst/go/deletenode.go
package deletenodebst

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val != key {
		if root.Val < key {
			root.Right = deleteNode(root.Right, key)
		} else {
			root.Left = deleteNode(root.Left, key)
		}

		return root
	}

	if root.Left == nil && root.Right == nil {
		return nil
	}

	if root.Left == nil {
		return root.Right
	}

	if root.Right == nil {
		return root.Left
	}

	newRoot := root.Right
	curr := root.Right

	for curr.Left != nil {
		curr = curr.Left
	}

	curr.Left = root.Left
	return newRoot
}

File: leetcode/450.delete-node-in-bst/cpp/deletenode.cpp

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *deleteNode(TreeNode *root, int key)
    {
        if (root == nullptr)
            return nullptr;

        if (root->val != key)
        {
            if (root->val < key)
                root->right = deleteNode(root->right, key);
            else
                root->left = deleteNode(root->left, key);
            return root;
        }

        if (root->left == nullptr)
            return root->right;
        if (root->right == nullptr)
            return root->left;

        auto newRoot = root->right;
        auto curr = root->right;

        while (curr->left != nullptr)
            curr = curr->left;

        curr->left = root->left;
        return newRoot;
    }
};
